pipeline/artifactory/artifactory.go

```diff
@@ -0,0 +1,348 @@
+	if instances == 0 {
+		return pipeline.Skip("artifactory section is not configured")
+	}
+
+	// Check if for every instance we have a the target,
+	// the username and a secret (password or api key)
+	// If not, we can skip this pipeline
+	for i := 0; i < instances; i++ {
+		if ctx.Config.Artifactories[i].Target == "" {
+			return pipeline.Skip(fmt.Sprintf("artifactory section is not configured properly (missing target in artifactory %d)", i))
+		}
+
+		if ctx.Config.Artifactories[i].Username == "" {
+			return pipeline.Skip(fmt.Sprintf("artifactory section is not configured properly (missing username in artifactory %d)", i))
+		}
+
+		envName := fmt.Sprintf("ARTIFACTORY_%d_SECRET", i)
+		if os.Getenv(envName) == "" {
+			return pipeline.Skip(fmt.Sprintf("missing secret for artifactory %d: %s", i, ctx.Config.Artifactories[i].Target))
+		}
+	}
+
+	return doRun(ctx)
+}
+
+func doRun(ctx *context.Context) error {
+	if !ctx.Publish {
+		return pipeline.Skip("--skip-publish is set")
+	}
+
+	// Loop over all builds, because we want to publish
+	// every build to Artifactory
+	for _, build := range ctx.Config.Builds {
+		if err := runPipeOnBuild(ctx, build); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// runPipeOnBuild runs the pipe for every configured build
+func runPipeOnBuild(ctx *context.Context, build config.Build) error {
+	sem := make(chan bool, ctx.Parallelism)
+	var g errgroup.Group
+
+	// Lets generate the build matrix, , because we want to publish
+	// every target to Artifactory
+	for _, target := range buildtarget.All(build) {
+		sem <- true
+		target := target
+		build := build
+		g.Go(func() error {
+			defer func() {
+				<-sem
+			}()
+
+			return doBuild(ctx, build, target)
+		})
+	}
+
+	return g.Wait()
+}
+
+// doBuild runs the pipe action of the current build and the current target
+// This is where the real action take place
+func doBuild(ctx *context.Context, build config.Build, target buildtarget.Target) (err error) {
+	binary, err := getBinaryForUploadPerBuild(ctx, target)
+	if err != nil {
+		return err
+	}
+
+	// Loop over all configured Artifactory instances
+
+	instances := len(ctx.Config.Artifactories)
+	for i := 0; i < instances; i++ {
+		artifactory := ctx.Config.Artifactories[i]
+		secret := os.Getenv(fmt.Sprintf("ARTIFACTORY_%d_SECRET", i))
+
+		// Generate name of target
+		uploadTarget, err := buildTargetName(ctx, artifactory, target)
+		if err != nil {
+			log.WithError(err).Error("Artifactory: Error while building the target name")
+			return errors.Wrap(err, "Artifactory: Error while building the target name")
+		}
+
+		// The upload url to Artifactory needs the binary name
+		// Here we add the binary to the target url
+		if !strings.HasPrefix(uploadTarget, "/") {
+			uploadTarget += "/"
+		}
+		uploadTarget += binary.Name
+
+		// Upload the binary to Artifactory
+		file, err := os.Open(binary.Path)
+		if err != nil {
+			return err
+		}
+		defer func() { err = file.Close() }()
```

Comment:
this err here would be missed since the method return nil, right?

Reply:
No, for this reason i introduced named returns in this function.
But i reverted it. The reason why i did this was because of gometalinter.
I ignored it now, because while fixing different gometalinter bugs, i introduced another bug which was fixed here. See https://github.com/goreleaser/goreleaser/pull/429/commits/37bc09f2bc4296ffb9d8e3e1e939529d8d4289a4